### PR TITLE
NX-OS static routes module

### PR DESCRIPTION
**Pull request by adharshsrivatsr**
##### SUMMARY
Resource module for nxos_static_routes in collections

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
nxos_static_routes

### CODE REVIEW

**Comment by NilashishC** [comment]
recheck

**Comment by trishnaguha** [comment]
@NilashishC @GomathiselviS Will you please review the nxos static route module? Looks like the `check` pipeline has run, it's waiting for review.

**Comment by NilashishC** [comment]
recheck

**Comment by NilashishC** [comment]
> @NilashishC @GomathiselviS Will you please review the nxos static route module? Looks like the `check` pipeline has run, it's waiting for review.

@trishnaguha - Zuul integration jobs haven't been enabled for the NX-OS collection, hence the check pipeline didn't run the tests. I have added two depends-on PRs that should take care of this.

**Comment by NilashishC** [comment]
recheck

**Comment by NilashishC** [comment]
CI failure for this target - https://object-storage-ca-ymq-1.vexxhost.net/v1/a0b4156a37f9453eb4ec7db5422272df/ansible_12/12/5fe22704007f57ba913bd9db328545ccdd0bfb78/check/ansible-test-network-integration-nxos-cli-python36/e0b1513/controller/ara-report/reports/e77041ac-c90b-4522-a5a9-3d2027dc4504.html

Looks like an import error in facts.py that is causing other RMs to fail as well.

**Comment by adharshsrivatsr** [comment]
> CI failure for this target - https://object-storage-ca-ymq-1.vexxhost.net/v1/a0b4156a37f9453eb4ec7db5422272df/ansible_12/12/5fe22704007f57ba913bd9db328545ccdd0bfb78/check/ansible-test-network-integration-nxos-cli-python36/e0b1513/controller/ara-report/reports/e77041ac-c90b-4522-a5a9-3d2027dc4504.html
> 
> Looks like an import error in facts.py that is causing other RMs to fail as well.

This is resolved.

**Comment by NilashishC** [comment]
recheck

**Comment by NilashishC** [comment]
recheck

**Comment by NilashishC** [comment]
@adharshsrivatsr static_routes is failing - https://object-storage-ca-ymq-1.vexxhost.net/v1/a0b4156a37f9453eb4ec7db5422272df/ansible_12/12/62ce2f0a5879f71a618897a6077656ce58461411/check/ansible-test-network-integration-nxos-cli-python36_4_of_4/b14d919/controller/ara-report/reports/1c1f8312-35e6-4f59-879d-41cccdb07461.html

**Comment by trishnaguha** [comment]
The assertion for idempotence test for overridden is failing https://object-storage-ca-ymq-1.vexxhost.net/v1/a0b4156a37f9453eb4ec7db5422272df/ansible_12/12/62ce2f0a5879f71a618897a6077656ce58461411/check/ansible-test-network-integration-nxos-cli-python36_4_of_4/b14d919/controller/ara-report/file/a77da7af-9c0d-4a81-8e0f-c631de9e274b/#line-53

File: plugins/modules/nxos_static_routes.py [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+    "metadata_version": "1.1",
+    "status": ["preview"],
+    "supported_by": "network",
+}
+
+DOCUMENTATION = """
+---
+module: nxos_static_routes
+version_added: "2.10"
+short_description: Configures and manages attributes of static routes on Cisco NX-OS platforms.
```
Comment (trishnaguha) [inline plugins/modules/nxos_static_routes.py:42]: Append the keyword "Static Routes Resource Module" at the end.

**Comment by NilashishC** [comment]
@adharshsrivatsr The `ip route 192.0.2.16/28 Ethernet1/2 192.0.2.45 vrf destinationVRF name overridden_route2"` command returns as a response of `ipv4 Nexthop: Interface eth1/2 is in switchport mode`. The interfaces of the NXOSv I5 CI appliance boot up as L2 ones. You need to explicitly run a `no switchport` on the Ethernet 1/2 interface in the populate task and again default them in remove task.

File: tests/integration/targets/nxos_static_routes/tests/cli/populate_config.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+- name: Populate config
```
Comment (NilashishC) [inline tests/integration/targets/nxos_static_routes/tests/cli/populate_config.yaml:2]: Please update this to:
```yaml
- name: Populate config
  cisco.nxos.nxos_config:
    lines:
      - "ip route 192.0.2.16/28 192.0.2.24 name new_route"
      - "ip route 192.0.2.80/28 192.0.2.26 tag 12"
      - "vrf context trial_vrf"
      - "  ip route 192.0.2.64/28 192.0.2.22 tag 4"
      - "  ip route 192.0.2.64/28 192.0.2.23 name merged_route 1"
      - "interface Ethernet1/2"
      - "  no switchport"
```

File: tests/integration/targets/nxos_static_routes/tests/cli/remove_config.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+---
+- name: Remove config
```
Comment (NilashishC) [inline tests/integration/targets/nxos_static_routes/tests/cli/remove_config.yaml:2]: Please update this to:
```
---
- name: Remove config
  cisco.nxos.nxos_config:
    lines:
      - "no ip route 192.0.2.16/28 192.0.2.24 name new_route"
      - "no ip route 192.0.2.80/28 192.0.2.26 tag 12"
      - "no ip route 192.0.2.44/30 192.0.2.55 tag 1 1"
      - "no ip route 192.0.2.16/28 192.0.2.23 name replaced_route1 3"
      - "no ip route 192.0.2.16/28 Ethernet1/2 192.0.2.45 vrf destinationVRF name replaced_route2"
      - "no ip route 192.0.2.36/30 192.0.2.32 name test_route1 tag 14"
      - "no ip route 192.0.2.36/30 192.0.2.48 name test_route2 2"
      - "no vrf context trial_vrf"
      - "default interface Ethernet1/2"
  ignore_errors: yes
  #error is thrown if we deleted trial_vrf when it is not present already. ignore_errors circumvents it
```

File: tests/integration/targets/nxos_static_routes/tests/cli/overridden.yml [hunk 4]
```diff
@@ -0,0 +1,58 @@
+---
+- debug:
+    msg: "Start nxos_static_routes overridden tests connection={{ ansible_connection }}"
+
```
Comment (NilashishC) [inline tests/integration/targets/nxos_static_routes/tests/cli/overridden.yml:4]: Please a remove_config call here, just to make sure we're not working on top of a spill over config.

File: tests/integration/targets/nxos_static_routes/tasks/cli.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+---
```
Comment (NilashishC) [inline tests/integration/targets/nxos_static_routes/tasks/cli.yaml:1]: Can you please update this to the content of https://raw.githubusercontent.com/ansible-collections/cisco.nxos/fc949445e6c4c91365c702ebf282c6394445011c/tests/integration/targets/nxos_lacp/tasks/cli.yaml?

File: tests/integration/targets/nxos_static_routes/defaults/main.yaml [hunk 6]
```diff
@@ -0,0 +1,2 @@
+---
+testcase: "*"
```
Comment (NilashishC) [inline tests/integration/targets/nxos_static_routes/defaults/main.yaml:2]:
```suggestion
testcase: '[^_].*'
```

Comment (NilashishC) [reply tests/integration/targets/nxos_static_routes/defaults/main.yaml:2]: Also, please add an underscore before remove_config and populate_config, so we don't run them as test files.

**Comment by NilashishC** [comment]
Also, please remove the hardcoded Ethernet1/2 to the any of the `nxos_intf` fact set in prepare_nxos_tests.

File: tests/.gitignore [hunk 7]
```diff
@@ -1 +1,2 @@
 output/
+integration/inventory.networking
```
Comment (NilashishC) [inline tests/.gitignore:2]: Please remove this file.

File: plugins/modules/nxos_static_routes.py [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+#               next_hops:
+#                 - forward_router_address: 2002:db8:12::1
+
+
+"""
+RETURN = """
+before:
+  description: The configuration prior to the model invocation.
+  returned: always
+  type: str
```
Comment (NilashishC) [inline plugins/modules/nxos_static_routes.py:450]:
```suggestion
  type: list
```

File: plugins/modules/nxos_static_routes.py [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+  sample: >
+    The configuration returned will always be in the same format
+     of the parameters above.
+after:
+  description: The resulting configuration model invocation.
+  returned: when changed
+  type: str
```
Comment (NilashishC) [inline plugins/modules/nxos_static_routes.py:457]:
```suggestion
  type: list
```

File: tests/integration/targets/nxos_static_routes/tasks/cli.yaml [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+- name: collect common test cases
+  find:
+    paths: "{{ role_path }}/tests/common"
```
Comment (NilashishC) [inline tests/integration/targets/nxos_static_routes/tasks/cli.yaml:4]: Please rename the tests/cli folder to tests/common

File: tests/integration/targets/nxos_static_routes/tests/cli/_remove_config.yaml [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    lines:
+      - "no ip route 192.0.2.16/28 192.0.2.24 name new_route"
+      - "no ip route 192.0.2.80/28 192.0.2.26 tag 12"
+      - "no ip route 192.0.2.44/30 192.0.2.55 tag 1 1"
+      - "no ip route 192.0.2.16/28 192.0.2.23 name replaced_route1 3"
+      - "no ip route 192.0.2.16/28 Ethernet1/2 192.0.2.45 vrf destinationVRF name replaced_route2"
+      - "no ip route 192.0.2.36/30 192.0.2.32 name test_route1 tag 14"
+      - "no ip route 192.0.2.36/30 192.0.2.48 name test_route2 2"
+      - "no vrf context trial_vrf"
+      - "default interface Ethernet1/2"
```
Comment (NilashishC) [inline tests/integration/targets/nxos_static_routes/tests/cli/_remove_config.yaml:13]:
```suggestion
      - "default interface {{ nxos_int2 }}"
```

File: tests/integration/targets/nxos_static_routes/tests/cli/overridden.yml [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+                  - dest: 192.0.2.16/28
+                    next_hops:
+                      - forward_router_address: 192.0.2.23
+                        route_name: overridden_route1
+                        admin_distance: 3
+
+                      - forward_router_address: 192.0.2.45
+                        route_name: overridden_route2
+                        dest_vrf: destinationVRF
+                        interface: Ethernet1/2
```
Comment (NilashishC) [inline tests/integration/targets/nxos_static_routes/tests/cli/overridden.yml:33]:
```suggestion
                        interface: "{{ nxos_int2 }}"
```

File: tests/integration/targets/nxos_static_routes/tests/cli/overridden.yml [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+      register: result
+
+    - assert:
+        that:
+          - "result.changed == true"
+          - "'vrf context trial_vrf' in result.commands"
+          - "'no ip route 192.0.2.64/28 192.0.2.22 tag 4' in result.commands"
+          - "'no ip route 192.0.2.64/28 192.0.2.23 name merged_route 1' in result.commands"
+          - "'ip route 192.0.2.16/28 192.0.2.23 name overridden_route1 3' in result.commands"
+          - "'ip route 192.0.2.16/28 Ethernet1/2 192.0.2.45 vrf destinationVRF name overridden_route2' in result.commands"
```
Comment (NilashishC) [inline tests/integration/targets/nxos_static_routes/tests/cli/overridden.yml:45]:
```suggestion
          - "'ip route 192.0.2.16/28 {{ nxos_int2 }} 192.0.2.45 vrf destinationVRF name overridden_route2' in result.commands"
```

File: tests/integration/targets/nxos_static_routes/tests/cli/_populate_config.yaml [hunk 14]
```diff
@@ -0,0 +1,11 @@
+---
+- name: Populate config
+  cisco.nxos.nxos_config:
+    lines:
+      - "ip route 192.0.2.16/28 192.0.2.24 name new_route"
+      - "ip route 192.0.2.80/28 192.0.2.26 tag 12"
+      - "vrf context trial_vrf"
+      - "  ip route 192.0.2.64/28 192.0.2.22 tag 4"
+      - "  ip route 192.0.2.64/28 192.0.2.23 name merged_route 1"
+      - "interface Ethernet1/2"
```
Comment (NilashishC) [inline tests/integration/targets/nxos_static_routes/tests/cli/_populate_config.yaml:10]:
```suggestion
      - "interface {{ nxos_int2 }}"
```

Comment (NilashishC) [reply tests/integration/targets/nxos_static_routes/tests/cli/_populate_config.yaml:10]: Please take care of this in all the places.

File: tests/integration/targets/nxos_static_routes/defaults/main.yaml (hunk 6 above)
Comment (NilashishC) [reply tests/integration/targets/nxos_static_routes/defaults/main.yaml:2]: This needs to be updated.

File: tests/integration/targets/nxos_static_routes/tasks/cli.yaml [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    patterns: "{{ testcase }}.yaml"
```
Comment (NilashishC) [inline tests/integration/targets/nxos_static_routes/tasks/cli.yaml:5]: Check for .yaml and .yml differences in test file names and make them consistent.

**Pull request merged**